app/src/main/java/com/example/faunaapp/view_model/LogInFragmentViewModel.java

```diff
@@ -16,48 +19,53 @@
 
 import org.jetbrains.annotations.NotNull;
 
-public class LogInFragmentViewModel extends ViewModel
-{
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.xml.transform.Result;
+
+public class LogInFragmentViewModel extends ViewModel {
     private MutableLiveData<String> errorMessage;
     private String errorConstructor;
     private Repository repository;
+    private ExecutorService executorService;
     private MutableLiveData<String> token;
 
+
     public LogInFragmentViewModel() {
         repository = Repository.getInstance();
         errorMessage = new MutableLiveData<>();
         token = new MediatorLiveData<>();
+        executorService = Executors.newFixedThreadPool(2);
     }
 
-    public LiveData<String> getErrorMessage(){
+    public LiveData<String> getErrorMessage() {
         return errorMessage;
     }
 
     public LiveData<String> getToken() {
-        token.postValue(repository.getToken());
+
         return token;
     }
 
-    public void logIn(String email, String password)
-    {
-        if(checkIfEmailAndPasswordExists(email, password))
-        {
-            repository.logIn(email, password);
+    public void logIn(String email, String password) {
+        if (checkIfEmailAndPasswordExists(email, password)) {
+            executorService.execute( () -> {
+                repository.logIn(email, password);
+                token.postValue(repository.getToken());
+            });
         }
     }
 
     private boolean checkIfEmailAndPasswordExists(String email, String password) {
         errorConstructor = "";
-        if(email.equals(""))
-        {
+        if (email.equals("")) {
             errorConstructor += "Please, provide your email\n";
         }
-        if(password.equals(""))
-        {
+        if (password.equals("")) {
             errorConstructor += "Please, provide your password";
         }
-        if(errorConstructor.equals(""))
-        {
+        if (errorConstructor.equals("")) {
```

Comment:
I do not think this is very expressive. It would be better to use your observable directly, and then consider returning false in either of your if statements above. Checking if the errorConstructor is empty is not very readable.

Reply:
Sure, I will change it